feat: Video editor supports transcripts [FC-0076]

ChrisChV opened the pull request:
<!--

Note: Please refer to the Support Development Guidelines on the wiki page to consider backporting to active releases:
https://openedx.atlassian.net/wiki/spaces/COMM/pages/4248436737/Support+Guidelines+for+active+releases

Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply.
You may link to information rather than copy it, but only if the link is publicly readable.
If the linked information must be private (because it contains secrets), clearly label the link as private.

-->

## Description

* Add error handler on save video to avoid creating sjson
* Support transcripts without `edx_video_id` in `definition_to_xml`
* When copying a video from a library to a course: Create a new `edx_video_id`
* Save transcripts as static assets in a video in a library when adding a new transcript.
* Delete transcripts as static assets in a video in a library when deleting transcripts.
* Support download transcript in a video in a library.
* Support replace transcript in a video in a library.
* Support updating transcripts in video in a library.
* Refactor the code of downloading YouTube transcripts to enable this feature in libraries.
* Support copy from a library to a course and a course to a library.
* Which edX user roles will this change impact? "Course Author"

## Supporting information

- Github link: https://github.com/openedx/frontend-app-authoring/issues/1352
- Used in: https://github.com/openedx/frontend-app-authoring/pull/1596
- Internal ticket: [FAL-3989](https://tasks.opencraft.com/browse/FAL-3989)

## Testing instructions

Follow the testing instructions at: https://github.com/openedx/frontend-app-authoring/pull/1596

## Deadline

No rush

## Other information

- [ ] TODO: The transcripts are not copied when you add a `library component` in a course. This will be fixed in https://github.com/openedx/modular-learning/issues/246

ChrisChV committed:
feat: Updates to support save video with transcript in library home

* Add error handler on save video to avoid create sjson
* Support transcripts without edx_video_id in definition_to_xml

openedx-webhooks commented:
Thanks for the pull request, @ChrisChV!


This repository is currently maintained by `@openedx/wg-maintenance-edx-platform`.


Once you've gone through the following steps feel free to tag them in a comment and let them know that your changes are ready for engineering review.

<details><summary>:radio_button: Get product approval</summary>

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.
</details>

<details><summary>:radio_button: Provide context</summary>

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads
</details>





<details><summary>:radio_button: Get a green build</summary>

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.
</details><details>


---

<details><summary>Where can I find more information?</summary>

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)
</details>

<details><summary>When can I expect my changes to be merged?</summary>

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*
</details>
<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

ChrisChV committed:
style: Fix lint

ChrisChV committed:
feat: Updated code to get english transcript from filename

ChrisChV committed:
style: Fix lint

ChrisChV committed:
feat: Upload transcript file as static asset in Learning Core in library components

ChrisChV committed:
style: Fix lint

ChrisChV committed:
feat: Updates to support download transcripts from youtube in vlibrary videos

ChrisChV committed:
style: Fix lint

ChrisChV committed:
test: Add test for download youtube transcripts in library content

ChrisChV committed:
style: Fix lint

ChrisChV committed:
feat: Support copy transcripts from a library

ChrisChV committed:
refactor: Allow use edx_video_id (edxval) in new runtime

This is used to be retroactive in copy-paste videos from Library to Course and Course to Library

ChrisChV committed:
refactor: Adds new edx_video_id when copy to course

ChrisChV committed:
refactor: Remove unnevessary code

ChrisChV committed:
feat: Support delete transcripts in library

ChrisChV committed:
style: Fix lint

ChrisChV committed:
Merge branch 'master' into chris/FAL-3989-video-transcripts

pomegranited commented on cms/djangoapps/contentstore/helpers.py:
```diff
@@ -616,6 +630,24 @@ def _import_file_into_course(
             content.thumbnail_location = thumbnail_location
         contentstore().save(content)
+        if usage_key.block_type == 'video':
+            # Adding transcripts to VAL using the nex edx_video_id
```
nit: typo
```suggestion
            # Adding transcripts to VAL using the new edx_video_id
```

ChrisChV replied:
Updated https://github.com/openedx/edx-platform/pull/36058/commits/590a5cda91f566801234e9f7f96b5f00dc553633

pomegranited commented on cms/djangoapps/contentstore/views/tests/test_transcripts.py:
```diff
@@ -702,6 +715,13 @@ def test_replace_transcript_success(self, edx_video_id):
+        response = self.replace_transcript(self.library_block.usage_key, self.youtube_id)
+
+        # Verify the response
+        self.assert_response(response, expected_status_code=200, expected_message='Success')
```
Could you also test that the transcript is replaced as expected?

ChrisChV replied:
Updated https://github.com/openedx/edx-platform/pull/36058/commits/0e5445055c64fc152eed35980c72365a21bd6e3b

pomegranited commented on cms/djangoapps/contentstore/views/transcripts_ajax.py:
```diff
@@ -81,13 +84,17 @@ def link_video_to_component(video_component, user):
         edx_video_id = create_external_video(display_name='external video')
+
+        if isinstance(video_component.usage_key, UsageKeyV2):
+            return edx_video_id
```
I don't understand why we're returning early here.. Could you add a comment to clarify?

ChrisChV replied:
Updated https://github.com/openedx/edx-platform/pull/36058/commits/e4f7c72496355e7eab29a5d0a7f9830808891c7b

bradenmacdonald replied:
Makes sense. But I wonder, should we still be calling `create_external_video` and returning an `edx_video_id` at all, if it's not going to be saved into the video block? Doesn't that create some stranded video data in VAL?

ChrisChV replied:
Updated https://github.com/openedx/edx-platform/pull/36058/files/34e5cbc493bf116a18d3e22ed139b3866e230161..46aa8ca7afe16896f37858c9290ea117adae8cdc

pomegranited commented on cms/djangoapps/contentstore/views/transcripts_ajax.py:
```diff
@@ -319,68 +342,83 @@ def check_transcripts(request):  # lint-amnesty, pylint: disable=too-many-statem
+        except NotFoundError:
+            log.debug("Can't find transcripts in storage for youtube id: %s", youtube_id)
+
+    #check youtube local and server transcripts for equality
```
nit: pylint should have caught this?
```suggestion
    # check youtube local and server transcripts for equality
```

ChrisChV replied:
Updated https://github.com/openedx/edx-platform/pull/36058/commits/590a5cda91f566801234e9f7f96b5f00dc553633

pomegranited commented on cms/djangoapps/contentstore/views/transcripts_ajax.py:
```diff
@@ -549,7 +588,13 @@ def choose_transcripts(request):
+            edx_video_id,
+            input_format,
+            transcript_content,
+            language_code='en',
```
I know you're dealing with the old transcripts code here, which is really convoluted.. but why is the language code hard-coded here (and below on `save_video_transcript`)? A comment would help :)

ChrisChV replied:
I'm sorry, but I'm not sure why it's hardcoded like that. This code is not used in video libraries (at least not as far as I've seen). I touched this code because I updated the `save_video_transcript` code.

pomegranited commented on cms/djangoapps/contentstore/views/transcripts_ajax.py:
```diff
@@ -456,7 +495,7 @@ def _validate_transcripts_data(request):
         raise TranscriptsRequestValidationException(_('Incoming video data is empty.'))
     try:
-        item = _get_item(request, data)
+        item, isLibraryContent = _get_item(request, data)
```
Is it hard to determine from the returned `item` whether it is a library block or not?

It's kind of icky that we're returning a tuple from `_get_item` just so this call can return it from `_validate_transcripts_data` so it can be passed into `_check_youtube_transcripts`..

Would be nicer if `_check_youtube_transcripts` examined the `item.location` to determine whether the `item` is a library block.

ChrisChV replied:
Updated https://github.com/openedx/edx-platform/pull/36058/commits/b478ac52927193e65559535e5b9e603adfca9a54

pomegranited commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -1952,3 +1952,6 @@ def import_blocks_create_task(library_key, course_key, use_course_key_as_block_i
              f"celery_task={result.id}")
     return import_task
+
+# To enable use content library permissions as public API
```
nit: grammar
```suggestion
# Allow content library permissions to be used in the public API
```

ChrisChV replied:
Updated https://github.com/openedx/edx-platform/pull/36058/commits/590a5cda91f566801234e9f7f96b5f00dc553633

pomegranited commented on xmodule/video_block/transcripts_utils.py:
```diff
@@ -508,6 +508,8 @@ def manage_video_subtitles_save(item, user, old_metadata=None, generate_translat
             except TranscriptException:
                 pass
+            except AttributeError:
+                pass
```
Why does this error need to be caught now? Seems a little dangerous.

ChrisChV replied:
Updated https://github.com/openedx/edx-platform/pull/36058/commits/5685f16389de256e1a4f5cb7010a7783ff7a96fd

pomegranited commented on xmodule/video_block/video_handlers.py:
```diff
@@ -497,6 +498,7 @@ def studio_transcript(self, request, dispatch):
         """
         _ = self.runtime.service(self, "i18n").ugettext
 
+        # pylint: disable=too-many-nested-blocks
```
nit: pylint is not wrong :/

Could this massive method be refactored into `_studio_transcript_upload`, `_studio_transcript_delete`, and `_studio_transcript_get` ? Unless that means we have to add more tests..

ChrisChV replied:
Updated https://github.com/openedx/edx-platform/pull/36058/commits/173a0afa08248c78b8a5e226352ecc1903e5b4e0

pomegranited replied:
That is so much nicer code to read, thank you!

pomegranited commented on xmodule/video_block/video_handlers.py:
```diff
@@ -562,22 +574,33 @@ def studio_transcript(self, request, dispatch):
-                    self.sub = ''  # pylint: disable=attribute-defined-outside-init
+                if isinstance(self.scope_ids.usage_id.context_key, LibraryLocatorV2):
+                    transcript_file_path = f"static/{self.transcripts.pop(language, None)}"
+                    lib_api.delete_library_block_static_asset_file(self.scope_ids.usage_id, transcript_file_path)
```
If `transcript_file_path == "static/None"`, I think `delete_library_block_static_asset_file` raises a validation error. By contrast, the `remove_subs_from_store` method just passes if the transcript file isn't found.

So maybe we should check for that here (at the risk of enraging pylint further..):
```suggestion
                    transcript_path = self.transcripts.pop(language, None)
                    if transcript_path:
                        lib_api.delete_library_block_static_asset_file(self.scope_ids.usage_id, f"static/{transcript_path}")
```

ChrisChV replied:
Updated https://github.com/openedx/edx-platform/pull/36058/commits/4532a2ae41af434503964bcde762c8e196936de5

pomegranited commented on cms/djangoapps/contentstore/helpers.py:
```diff
@@ -10,6 +10,7 @@
 import re
```
I'm seeing a bug when I sync a LibraryBlock video with transcripts from an upstream video.

Steps to reproduce:

1. Create a library video with transcripts (here, I imported them from the [example youtube video](https://www.youtube.com/watch?v=dGby9BH9bMc)).
2. Publish the library video.
3. Copy it to the clipboard.
4. Paste into a course.
    Note that the transcripts are displaying fine here.
5. Re-edit the library video, and replace a transcript. (Here, I replaced the English one, I don't know if replacing others causes the same issue).
6. Return to the course LibraryBlock, and refresh to see the "updates available" button. Click it.
   Note that the upstream video preview shows its transcripts fine, but the downstream (course) video preview doesn't show its transcripts anymore.
7. Accept changes.
   Note that the course video no longer shows its transcripts, but if you edit it, you can see they're still there.

https://github.com/user-attachments/assets/887e6d96-ead1-4586-a19b-41bd69cde7e6

ChrisChV replied:
I think this is related to https://github.com/openedx/modular-learning/issues/246

pomegranited replied:
@ChrisChV That could very well be.. however I don't think it's resolved by @DanielVZ96 's https://github.com/openedx/edx-platform/pull/36173, but it's also possible that I didn't merge conflicts accurately. cf [my merged branch](https://github.com/openedx/edx-platform/compare/master...open-craft:edx-platform:jill/FAL-3989+FAL-4027).

ChrisChV replied:
@pomegranited To be safe, I will wait until https://github.com/openedx/edx-platform/pull/36173 is ready to fix this bug.

pomegranited replied:
No worries @ChrisChV , thank you for keeping an eye on this issue.

bradenmacdonald replied:
Now that that other PR is merged, any update on this?

ChrisChV replied:
@bradenmacdonald @pomegranited I fixed this on https://github.com/openedx/edx-platform/pull/36058/commits/0745f0917a5e76c3578777e9cb9416885d47aea2. I need to add tests, but you can test the fix

pomegranited replied:
Hi @ChrisChV -- your fix is an improvement, but now I'm seeing a slightly different bug in the "updates available" preview.

Steps to reproduce:

1. Create a library video with one or more transcripts (here, I used your mp4 [video link](https://www.sample-videos.com/video321/mp4/720/big_buck_bunny_720p_2mb.mp4), but it happens with the imported transcripts from the [example youtube video](https://www.youtube.com/watch?v=dGby9BH9bMc) too).
1. Publish the library video.
1. Copy it to the clipboard.
1.  Paste into a course.
    Note that the transcripts are displaying fine here (despite the message about needing to upload transcript files??)
1. Re-edit the library video, and replace a transcript. Wait for it to process fully before hitting Save.
1. Publish the library video again.
1. Return to the course LibraryBlock, and refresh to see the "updates available" button. Click it.
    Bug: the downstream (course) video preview **shows the updated English transcript from upstream**, not the actual transcript that's on the downstream block.
1. Accept changes.
   Sync works fine (sorry, screencapture below doesn't show this step, but it does work :) )

https://github.com/user-attachments/assets/dbba8b3a-c363-41ee-a55c-be1f9b62c53a

ChrisChV replied:
Thanks @pomegranited! Do you think we can move that issue to another ticket? I've already spent a lot of budget on this task, and it's getting much bigger every time :smiley:

pomegranited replied:
@ChrisChV Yes I think so -- the syncing works, it's just the preview that's oddly borked. And AFAIK, syncing was out of scope here anyway :)

ChrisChV replied:
Thanks! I created the issue: https://github.com/openedx/frontend-app-authoring/issues/1682

pomegranited reviewed (changes requested):
Hi @ChrisChV , this is working well for the most part, good job dealing with the old transcript code!

But I found a bug with the upstream/downstream syncing, and left a few nits/change requests too.

ChrisChV committed:
style: Nits on the code

ChrisChV committed:
refactor: Update replace_transcript to verify transcript

ChrisChV committed:
refactor: Verify LibraryLocatorV2 in manage_video_subtitles_save to avoid raise AttributeError

ChrisChV committed:
style: Add comment in transcripts_ajax.py

ChrisChV committed:
refactor: _get_item to avoid return isLibraryContent

ChrisChV committed:
refactor: studio_transcript to separate methods in separated functions

ChrisChV committed:
refactor: Update code in video_handlers to avoid "static/None"

pomegranited reviewed (approved):
:+1: Thank you for making those changes @ChrisChV ! Code looks and works great.

- [x] I tested this using the testing instructions from https://github.com/openedx/frontend-app-authoring/pull/1596.
   I also tested "duplicating" video blocks with transcripts in courses, and they worked too.
- [x] I read through the code
- [x] I checked for accessibility issues by using my keyboard to navigate
- [x] Includes documentation -- good code comments
- [x] <strike>User-facing strings are extracted for translation</strike> N/A

ChrisChV committed:
style: Fix nit

DanielVZ96 commented on cms/djangoapps/contentstore/helpers.py:
```diff
@@ -299,13 +302,21 @@ def import_staged_content_from_user_clipboard(parent_key: UsageKey, request) ->
         )
 
+        usage_key = new_xblock.scope_ids.usage_id
+        if usage_key.block_type == 'video':
```
This overlaps a bit with the changes at https://github.com/openedx/edx-platform/pull/36173. Since there are some refactors to these functions here. I'll wait for the merge of this PR to update https://github.com/openedx/edx-platform/pull/36173 with your changes

bradenmacdonald replied:
Since that PR merged first, does this one need to be updated?

DanielVZ96 reviewed (approved):
:+1:

- [x] I tested this
- [x] I read through the code
- [x] I checked for accessibility issues

bradenmacdonald commented on cms/djangoapps/contentstore/helpers.py:
```diff
@@ -299,13 +302,21 @@ def import_staged_content_from_user_clipboard(parent_key: UsageKey, request) ->
             tags=user_clipboard.content.tags,
         )
 
+        usage_key = new_xblock.scope_ids.usage_id
```
```suggestion
        usage_key = new_xblock.usage_key
```
If you want, there is now a simpler way to get this :)

ChrisChV replied:
Updated https://github.com/openedx/edx-platform/pull/36058/commits/a254fe4a5672e796baf46c3caa40d569a0203133

bradenmacdonald commented on cms/djangoapps/contentstore/helpers.py:
```diff
@@ -616,6 +630,24 @@ def _import_file_into_course(
+                        'language_code': language_code
+                    },
+                    file_data=ContentFile(sjson_subs),
+                )
```
This new code does not seem to match the docstring of the function: "Import a single staged static asset file into the course, unless it already exists." It also doesn't use `staged_content_id` nor `file_data_obj`.

I think the code is fine but it should be moved out of `_import_file_into_course` and into a new helper function like `_import_transcripts`

ChrisChV replied:
Updated https://github.com/openedx/edx-platform/pull/36058/commits/34e5cbc493bf116a18d3e22ed139b3866e230161

bradenmacdonald reviewed (commented):
This is a big PR so I haven't finished reviewing yet, but here a couple questions so far.

bradenmacdonald commented on cms/djangoapps/contentstore/views/transcripts_ajax.py:
```diff
@@ -118,6 +128,21 @@ def save_video_transcript(edx_video_id, input_format, transcript_content, langua
+                output_format=Transcript.SRT
+            ).encode()
+
+            filename = f"static/{edx_video_id}-{language_code}.srt"
```
Do we need to put the `edx_video_id` in the filename? Because `transcript-{language_code}.srt` would be a much nicer name.

ChrisChV replied:
Updated https://github.com/openedx/edx-platform/pull/36058/commits/46aa8ca7afe16896f37858c9290ea117adae8cdc

bradenmacdonald commented on cms/djangoapps/contentstore/views/transcripts_ajax.py:
```diff
@@ -643,17 +705,26 @@ def replace_transcripts(request):
+                request.user,
+                lib_api.lib_permissions.CAN_EDIT_THIS_CONTENT_LIBRARY
+            )
+            return xblock_api.load_block(usage_key, request.user)
```
There is already a public API for this:
```suggestion
            return xblock_api.load_block(usage_key, request.user, check_permission=CheckPerm.CAN_EDIT)
```

ChrisChV replied:
Updated: https://github.com/openedx/edx-platform/pull/36058/commits/5bb426f505c1f0ba3b862f992a62dd2ac78091bd

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -1952,3 +1952,6 @@ def import_blocks_create_task(library_key, course_key, use_course_key_as_block_i
     return import_task
+
+# Allow content library permissions to be used in the public API
+lib_permissions = permissions
```
See comment above - we shouldn't expose the permissions as part of the API because they're considered internal to the content libraries feature. But you _can_ use the `xblock_api.load_block(..., check_permission=...)` API to require permissions when you load the block.

bradenmacdonald commented on xmodule/tests/test_video.py:
```diff
@@ -741,6 +741,48 @@ def test_export_to_xml(self, mock_val_api):
 
+    def test_export_to_xml_without_video_id(self):
+        """
+        Test that we write the correct XML without video_id on export.
```
It's a bit unclear from this wording - is this testing "that we write the correct XML for a block that doesn't have a video_id" or is this testing that "the correct XML should not include video_id when exported" ?

ChrisChV replied:
Updated https://github.com/openedx/edx-platform/pull/36058/commits/5dda777a2fce34cf9c4869024a5da210b5a7e3ed

bradenmacdonald commented on xmodule/video_block/video_handlers.py:
```diff
@@ -495,118 +495,161 @@ def studio_transcript(self, request, dispatch):
-                    delete_video_transcript(video_id=edx_video_id, language_code=language)
+    def _studio_transcript_upload(self, request):
+        """
+        Upload transcript. Usedn in "POST" method in `studio_transcript`
```
```suggestion
        Upload transcript. Used in "POST" method in `studio_transcript`
```

ChrisChV replied:
Updated https://github.com/openedx/edx-platform/pull/36058/commits/53f42fa8c72333217d99f1c9917eaeed2ca8af54

bradenmacdonald commented on xmodule/video_block/video_handlers.py:
```diff
@@ -495,118 +495,161 @@ def studio_transcript(self, request, dispatch):
-                    remove_subs_from_store(self.transcripts.pop(language, None), self, language)
+    def _studio_transcript_delete(self, request):
+        """
+        Delete transcript. Usedn in "DELETE" method in `studio_transcript`
```
```suggestion
        Delete transcript. Used in "DELETE" method in `studio_transcript`
```

ChrisChV replied:
Updated https://github.com/openedx/edx-platform/pull/36058/commits/53f42fa8c72333217d99f1c9917eaeed2ca8af54

bradenmacdonald commented on xmodule/video_block/video_handlers.py:
```diff
@@ -495,118 +495,161 @@ def studio_transcript(self, request, dispatch):
 
+    def _studio_transcript_get(self, request):
+        """
+        Get transcript. Usedn in "GET" method in `studio_transcript`
```
```suggestion
        Get transcript. Used in "GET" method in `studio_transcript`
```

ChrisChV replied:
Updated https://github.com/openedx/edx-platform/pull/36058/commits/53f42fa8c72333217d99f1c9917eaeed2ca8af54

bradenmacdonald reviewed (commented):
@ChrisChV Nice work on a very complex and ugly part of the code 👏🏻. I have a few small changes to request but I think this is just about good to go.

ChrisChV committed:
chore: Fix merge conflicts

ChrisChV committed:
refactor: Use usage_key in some places of the code

ChrisChV committed:
refactor: _import_transcripts created in helpers.py

ChrisChV committed:
refactor: save_video_transcript_in_learning_core created

save_video_transcript_in_learning_core created in transcript_ajax to avoid create an edx_video_id for library videos

ChrisChV committed:
refactor: _studio_transcript_upload in video_handlers

* _save_transcript_field created to save the `transcripts` Field
* _studio_transcript_upload updated to avoid create edx_video_id for libraries
* _studio_transcript_upload updated to avoid add transcripts to VAL for libraries

ChrisChV committed:
refactor: transcripts_ajax to avoid use edx_video_id in filenames in libraries

ChrisChV committed:
refactor: Update _get_item in transcripts_ajax to use check_permission of load_block

ChrisChV committed:
refactor: Delete lib_permissions from content_libraries/api.py

ChrisChV committed:
style: Fix typos in comments in video_handlers.py

ChrisChV committed:
style: Update comment of test_export_to_xml_without_video_id

ChrisChV committed:
fix: Fix tests and lint

ChrisChV committed:
fix: Sync transcripts in video blocks from library to course

ChrisChV committed:
test: Tests for sync video block

ChrisChV committed:
style: Fix lint

bradenmacdonald commented on xmodule/video_block/video_handlers.py:
```diff
@@ -527,29 +541,21 @@ def _studio_transcript_upload(self, request):
-                # Back-populate the video ID for an external video.
-                # pylint: disable=attribute-defined-outside-init
-                self.edx_video_id = edx_video_id = create_external_video(display_name='external video')
+            isLibrary = isinstance(self.usage_key.context_key, LibraryLocatorV2)
```
Nit: python vars should be `is_library` not `isLibrary`

ChrisChV replied:
Updated https://github.com/openedx/edx-platform/pull/36058/commits/6903be2b674410d06f07572f7acc6c37b3f23482

bradenmacdonald commented on cms/djangoapps/contentstore/views/transcripts_ajax.py:
```diff
@@ -87,6 +91,45 @@ def link_video_to_component(video_component, user):
+    language_code
+):
+    """
+    Saves a video transcript to the learning core
```
Might be worth adding a comment/docstring that SJSON is an edx-specific format that we're trying to move away from, so all new (Learning Core) stuff should only use the standard `.srt` format for subtitles.

ChrisChV replied:
Added in as comment https://github.com/openedx/edx-platform/pull/36058/commits/6903be2b674410d06f07572f7acc6c37b3f23482

bradenmacdonald commented on xmodule/video_block/video_handlers.py:
```diff
@@ -495,118 +495,179 @@ def studio_transcript(self, request, dispatch):
+        if isinstance(self.usage_key.context_key, LibraryLocatorV2):
+            transcript_name = self.transcripts.pop(language, None)
+            if transcript_name:
+                lib_api.delete_library_block_static_asset_file(
```
Note (please add as a comment): In the future, we need a proper XBlock API like `self.static_assets.delete(...)` instead of coding these runtime-specific/library-specific APIs.

ChrisChV replied:
Added https://github.com/openedx/edx-platform/pull/36058/commits/6903be2b674410d06f07572f7acc6c37b3f23482

bradenmacdonald reviewed (approved):
Code looks good to me - thanks for those updates!

@pomegranited can you please test this carefully again before merging? Or let me know if you want help.

ChrisChV committed:
style: Nits and comments in the code

ChrisChV committed:
style: Fix lint

pomegranited reviewed (approved):
@ChrisChV This is working great, except for one weird issue still in the upstream/downstream preview. So I'm approving, in case this can be fixed in a follow-up PR?

- [x] I tested this using the test instructions from https://github.com/openedx/frontend-app-authoring/pull/1596, and also by syncing transcript changes from library blocks to downstream course blocks.
- [x] I read through the code
- [x] I checked for accessibility issues by using my keyboard to navigate
- [x] Includes documentation
- [x] User-facing strings are extracted for translation

ChrisChV committed:
Merge branch 'master' into chris/FAL-3989-video-transcripts

kdmccormick commented on cms/lib/xblock/upstream_sync.py:
```diff
@@ -297,7 +297,11 @@ def _update_non_customizable_fields(*, upstream: XBlock, downstream: XBlock) ->
     """
     syncable_fields = _get_synchronizable_fields(upstream, downstream)
     customizable_fields = set(downstream.get_customizable_fields().keys())
+    isVideoBlock = downstream.usage_key.block_type == "video"
     for field_name in syncable_fields - customizable_fields:
+        if isVideoBlock and field_name == 'edx_video_id':
+            # Avoid overwriting edx_video_id between blocks
+            continue
```
@ChrisChV @pomegranited With all due appreciation for the VideoBlock nonsense you've had to grapple with to support transcripts, it truly pains me to see such block-type-specific code in this highly generalized function 😓 Upstream syncing is complex but nature, but adding special cases like this move it from being "complex" to being "complicated".

Is there any other way we can ensure that edx_video_id isn't synced? Would it not work to put it in the dict returned by `get_customizable_fields`, as @pomegranited did for `graceperiod`, `grading_method`, etc.?

bradenmacdonald replied:
+1, good point

Though the two root problems here are (1) that there is no standard way to associate a block with external data that's not in an XBlock field (e.g. edx_video_id points to VAL or ORA2 has tons of Django models holding its data) and (2) that there is no proper XBlock API for dealing with static files (e.g. transcripts)
